Add support for Recordings Edl

rbarthen opened the pull request:
ArgusTV PVR has been missing EDL support for recordings since version 17.  I don't normally program C++, but I've been missing the feature enough to try and contribute.  I've borrowed most of the code needed from the pvr.wmc addon to support EDL.  I included the entire Utils Namespace from pvr.wmc, but not all of it is needed for Edl support.  I left the other methods for potential future use.

rbarthen committed:
Add support for Recordings Edl
with recommended amendments

phunkyfish commented:
Thanks for contributing!

The split function is actually available in the binary addon API. You can see an example of it being used in `argustvrpc.cpp`.

So I reckon you can drop the Utils completely, giving you a much shorter PR. Also, please create the PR for Nexus also (that's like master for binary addons). 

Last thing to do then is to increment the middle number in `addon.xml.in` and add an entry to the changelog file.

phunkyfish commented on src/pvrclient-argustv.h:
```diff
@@ -122,6 +126,7 @@ class cPVRClientArgusTV : public kodi::addon::CInstancePVRClient
   std::string m_sBackendVersion;
   time_t m_BackendUTCoffset = 0;
   time_t m_BackendTime = 0;
+  std::string _streamFileName; // the name of the stream file
```
`m_streamFilename`

Note that case change for the `n` also. But I'm not sure you need a member for this variable. It should be ok to just declare it in the function where it's used. Unless there is another reason to have it as a member?

phunkyfish commented on src/utils.cpp:
```diff
@@ -18,6 +18,11 @@
 #include <kodi/General.h>
 #include <string>
 
+
+#include <kodi/Filesystem.h>
+#include <stdio.h>
+
+
```
Likely the changes in this file are no longer needed.

phunkyfish commented on src/utils.cpp:
```diff
@@ -117,6 +122,203 @@ std::string b64_encode(unsigned char const* in, unsigned int in_len, bool urlEnc
+  if (found != std::string::npos)
+    return path.substr(0, found);
+  else
+    return path;
+}
+
+bool ReadFileContents(std::string const& strFileName, std::string& strContent)
+{
+  kodi::vfs::CFile fileHandle;
+  if (fileHandle.OpenFile(strFileName))
+  {
+    std::string buffer;
+    while (fileHandle.ReadLine(buffer))
+      strContent.append(buffer);
+    return true;
+  }
+  return false;
+}
+
+bool WriteFileContents(std::string const& strFileName, const std::string& strContent)
+{
+  kodi::vfs::CFile fileHandle;
+  if (fileHandle.OpenFileForWrite(strFileName, true))
+  {
+    int rc = fileHandle.Write(strContent.c_str(), strContent.length());
+    if (rc)
+    {
+      kodi::Log(ADDON_LOG_DEBUG, "wrote file %s", strFileName.c_str());
+    }
+    else
+    {
+      kodi::Log(ADDON_LOG_ERROR, "can not write to %s", strFileName.c_str());
+    }
+    return rc >= 0;
+  }
+  return false;
+}
+
+} /* namespace Utils */
+
```
Likely the changes in this file are no longer needed

phunkyfish commented on src/utils.h:
```diff
@@ -11,6 +11,9 @@
 #include <json/json.h>
 #include <string>
 
+#include "stdio.h"
+#include <vector>
+
```
Likely the changes in this file are no longer needed

phunkyfish commented on src/utils.h:
```diff
@@ -32,6 +35,23 @@ namespace BASE64
 std::string b64_encode(unsigned char const* in, unsigned int in_len, bool urlEncode);
 }
 
+namespace Utils
+{
+
+std::vector<std::string> Split(const std::string& input,
+                               const std::string& delimiter,
+                               unsigned int iMaxStrings = 0);
+bool Str2Bool(const std::string& str);
+std::string Format(const char* fmt, ...);
+std::string FormatV(const char* fmt, va_list args);
+bool EndsWith(std::string const& fullString, std::string const& ending);
+bool StartsWith(std::string const& fullString, std::string const& starting);
+std::string GetDirectoryPath(std::string const& path);
+bool ReadFileContents(std::string const& strFileName, std::string& strResult);
+bool WriteFileContents(std::string const& strFileName, const std::string& strContent);
+
+} /* namespace Utils */
+
```
Likely the changes in this file are no longer needed

phunkyfish commented on src/pvrclient-argustv.cpp:
```diff
@@ -867,6 +868,62 @@ PVR_ERROR cPVRClientArgusTV::SetRecordingPlayCount(const kodi::addon::PVRRecordi
+PVR_ERROR cPVRClientArgusTV::GetRecordingEdl(const kodi::addon::PVRRecording& recording,
+                                             std::vector<kodi::addon::PVREDLEntry>& edl)
+{
+  if (!FindRecEntry(recording.GetRecordingId(), _streamFileName))
```
streamFilename is only used in this function so why not declare it here instead of in the header?

`std::string streamFilename; `

phunkyfish commented on src/pvrclient-argustv.cpp:
```diff
@@ -867,6 +868,62 @@ PVR_ERROR cPVRClientArgusTV::SetRecordingPlayCount(const kodi::addon::PVRRecordi
+  if (!FindRecEntry(recording.GetRecordingId(), _streamFileName))
+    return PVR_ERROR_SERVER_ERROR;
+
+  if (!_streamFileName.empty()) // read the edl for the current stream file
```
`streamFilename`

phunkyfish commented on src/pvrclient-argustv.cpp:
```diff
@@ -867,6 +868,62 @@ PVR_ERROR cPVRClientArgusTV::SetRecordingPlayCount(const kodi::addon::PVRRecordi
+  if (!_streamFileName.empty()) // read the edl for the current stream file
+  {
+    // see if edl file for currently streaming recording exists
+    std::string theEdlFile = _streamFileName;
```
`streamFilename`

phunkyfish reviewed (changes requested):
Not sure you need a member variable for `streamFilename`

phunkyfish commented on src/pvrclient-argustv.cpp:
```diff
@@ -872,13 +873,14 @@ PVR_ERROR cPVRClientArgusTV::SetRecordingPlayCount(const kodi::addon::PVRRecordi
                                              std::vector<kodi::addon::PVREDLEntry>& edl)
 {
-  if (!FindRecEntry(recording.GetRecordingId(), _streamFileName))
+  std::string m_streamFileName; // the name of the stream file
```
It’s no longer a member variable so should just be called `streamFilename`

phunkyfish commented on src/pvrclient-argustv.cpp:
```diff
@@ -872,13 +873,14 @@ PVR_ERROR cPVRClientArgusTV::SetRecordingPlayCount(const kodi::addon::PVRRecordi
     return PVR_ERROR_SERVER_ERROR;
 
-  if (!_streamFileName.empty()) // read the edl for the current stream file
+  if (!m_streamFileName.empty()) // read the edl for the current stream file
```
`streamFilename`

phunkyfish commented on src/pvrclient-argustv.cpp:
```diff
@@ -872,13 +873,14 @@ PVR_ERROR cPVRClientArgusTV::SetRecordingPlayCount(const kodi::addon::PVRRecordi
   {
     // see if edl file for currently streaming recording exists
-    std::string theEdlFile = _streamFileName;
+    std::string theEdlFile = m_streamFileName;
```
`streamFilename`

phunkyfish reviewed (changes requested):
Please also squash the 3 commits into one.

phunkyfish commented on src/pvrclient-argustv.h:
```diff
@@ -122,7 +126,6 @@ class cPVRClientArgusTV : public kodi::addon::CInstancePVRClient
   std::string m_sBackendVersion;
   time_t m_BackendUTCoffset = 0;
   time_t m_BackendTime = 0;
-
```
No need to remove this line

rbarthen commented:
I'm not sure if I did the squashing correctly.  The updates should have the following corrections:
- minor version updated
- changelog updated
- using addon library for split
- removed utils namespace
- stream file naming corrected
- added back removed line return

phunkyfish commented on src/pvrclient-argustv.h:
```diff
@@ -122,7 +126,7 @@ class cPVRClientArgusTV : public kodi::addon::CInstancePVRClient
   time_t m_BackendUTCoffset = 0;
   time_t m_BackendTime = 0;
-
+  
```
Remove spaces

phunkyfish commented:
> I'm not sure if I did the squashing correctly. The updates should have the following corrections:
> 
> * minor version updated
> * changelog updated
> * using addon library for split
> * removed utils namespace
> * stream file naming corrected
> * added back removed line return

Once you have squashed it you should only have a single commit. Then you would force push to your PR branch. What I recommend you do is this.

1. `git checkout -b Matrix-bk` (this will just give you a backup branch in case you need it)
2. `git checkout Matrix` (Switch back to working branch)
3. `git reset HEAD~1` (this will undo your latest commit and put it back to an uncommitted change)
4. Keep running `git reset HEAD~1` until you are back to your origin commit.
5. Once you are back to your original commit (you can check that with `git log --oneline` then:
6. `git add -u`
7. `git commit --amend` (this will add all the changes to your original commit)

Note that `git add -u` and `git commit --amend` is a common pattern to combine into the last commit.

rbarthen commented:
@phunkyfish ,
Apologies for my intermediate knowledge of git. I tried your instructions, but it didn't seem to work because I mistakenly squashed the original PR commit and could not move the HEAD back to it. I checked out that hash for the PR comment, but it resulted in a detached head. I manually merged the latest updates into that detached head and committed with an amend, but it wouldn't push without forcing. I now only see the one commit with a new hash. Hopefully that put things back on track.

phunkyfish commented:
> @phunkyfish ,
> Apologies for my intermediate knowledge of git. I tried your instructions, but it didn't seem to work because I mistakenly squashed the original PR commit and could not move the HEAD back to it. I checked out that hash for the PR comment, but it resulted in a detached head. I manually merged the latest updates into that detached head and committed with an amend, but it wouldn't push without forcing. I now only see the one commit with a new hash. Hopefully that put things back on track.

This is correct now. You do need to force push to after squashing as you are overwriting the commit history. Well done for sticking with it!

phunkyfish reviewed (approved):
Thanks for the contribution!,,

phunkyfish commented:
Now all you need to do is create a similar PR for the Nexus branch. Nexus is our development branch and everything needs to be merged there first.

When you checkout this branch remember to create a working copy first like this:

- `git checkout Nexus` (Get the main development branch)
- `git pull` (Make sure it’s up to date)
- `git checkout -b recording-delivery-matrix` (this is the dev branch you add commits too)

We work like this so that if the main branch changes in the meantime you can still rebase easily. 

All you need to do is cherry-pick the commit (using the hash) over and amend the addon.xml.in and the change log.

EDIT: please wait a few minutes before adding the PR for Nexus. There is another PR that need to be merged before yours is created.

phunkyfish commented:
@rbarthen the Nexus branch has now been updated so you can create the other PR for there now. I’m not sure how your git repos are setup but if you need any help just ask.

phunkyfish commented:
> @rbarthen the Nexus branch has now been updated so you can create the other PR for there now. I’m not sure how your git repos are setup but if you need any help just ask.

@rbarthen I'll create the Nexus PR for you. There is an API change so the Nexus add-on needs to be released anyway and it's very easy to cheery-pick your change over. Thanks again and well done on your first contribution.

rbarthen commented:
@phunkyfish ,
Thanks for handling Nexus.  I had started the code change, but got distracted with New Years Eve events.
